runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/BundleProcessor.java

```diff
@@ -0,0 +1,151 @@
+package org.apache.beam.runners.fnexecution.control;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.runners.fnexecution.data.FnDataService;
+import org.apache.beam.runners.fnexecution.data.RemoteInputDestination;
+import org.apache.beam.runners.fnexecution.state.StateDelegator;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandler;
+import org.apache.beam.sdk.fn.data.CloseableFnDataReceiver;
+import org.apache.beam.sdk.fn.data.InboundDataClient;
+import org.apache.beam.sdk.fn.data.LogicalEndpoint;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A processor capable of creating bundles for some registered
+ * {@link BeamFnApi.ProcessBundleDescriptor}.
+ */
+public class BundleProcessor<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(BundleProcessor.class);
+
+  private final BeamFnApi.ProcessBundleDescriptor processBundleDescriptor;
+  private final CompletionStage<BeamFnApi.RegisterResponse> registrationFuture;
+  private final RemoteInputDestination<WindowedValue<T>> remoteInput;
+  private final StateDelegator stateDelegator;
+  private InstructionRequestHandler fnApiControlClient;
+  private FnDataService fnApiDataService;
+  private SdkHarnessClient.IdGenerator idGenerator;
+
+  BundleProcessor(
+      InstructionRequestHandler fnApiControlClient,
+      FnDataService fnApiDataService,
+      SdkHarnessClient.IdGenerator idGenerator,
+      BeamFnApi.ProcessBundleDescriptor processBundleDescriptor,
+      CompletionStage<BeamFnApi.RegisterResponse> registrationFuture,
+      RemoteInputDestination<WindowedValue<T>> remoteInput,
+      StateDelegator stateDelegator) {
+    this.processBundleDescriptor = processBundleDescriptor;
+    this.registrationFuture = registrationFuture;
+    this.remoteInput = remoteInput;
+    this.stateDelegator = stateDelegator;
+    this.fnApiControlClient = fnApiControlClient;
+    this.fnApiDataService = fnApiDataService;
+    this.idGenerator = idGenerator;
+  }
+
+  public CompletionStage<BeamFnApi.RegisterResponse> getRegistrationFuture() {
```

Comment:
We probably want to expose something simpler (`isRegistrationComplete() throws Exception` or whatever).

This may not even need to be exposed outside of the package at all, and the `SdkHarnessClient` could handle awaiting the response and making sure that it was successful or not.

It might be worthwhile to have the `BundleProcessor` update its state if the stage completes exceptionally.

Reply:
I mean it really depends on the intended usage pattern.  getRegistrationFuture would make sense if you want to attach an async callback once registration completes. 

However, since all code that currently calls it would benefit from your simplification, I'm into it.